docs/src/references/operation-workflow.md

```diff
@@ -0,0 +1,376 @@
+On reception of a software update request from the cloud operator,
+the `tedge-mapper` creates a fresh new topic for this command,
+say `te/device/main///cmd/software-update/req-123` for the 123<sup>rd</sup> request.
+On this topic, a first retained messages is published to describe the operator expectations for the software updates.
+
+```sh te2mqtt
+tedge mqtt pub -r 'te/device/main///cmd/software-update/req-123' '{
+    "status": "init",
+    "modules": [
+        {
+            "type": "apt",
+            "name": "collectd",
+            "version": "5.7",
+            "action": "install"
+        }
+    ]
+}' 
+```
+
+Then, the `tedge-agent` and possibly other software components take charge of the command, making it advance to some final state,
+publishing all the successive states as retained messages on the command topic.
+
+Eventually, the `tedge-mapper` will have to clean the command topic with an empty retained message: 
+
+```sh te2mqtt
+tedge mqtt pub -r 'te/device/main///cmd/software-update/req-123' ''
+```
+
+## MQTT-Driven Workflows
+
+Operations that require coordination among several software components are managed using *MQTT-driven workflows*.
+
+The core idea is to expose over MQTT the different states a specific operation request might go through;
+so independent sub-systems can observe the progress of the request and act accordingly to their role.
+
+- A specific topic is attached to each command under-execution.
+  - This topic is specific to the target of the command, the requested operation and the request instance.
+  - e.g. `te/device/child-xyz///cmd/configuration-update/req-123`
+- The messages published over this topic represent the current state of the command.
+  - Such a message states where the command progression is and gives all the required information to proceed.
+  - e.g. `{ "status": "Requested", "target": "mosquitto", "url": "https://..." }`
+- The state messages are published as retained.
+  - They capture the latest state of the operation request.
+  - Till some change occurs, this latest state is dispatched to any participant on reconnect.
+- Several participants act in concert to move forward the command execution.
+  - The participants observe the progress of all the operations they are interested in.
+  - They watch for the specific states they are responsible in moving forward.
+  - When a step is performed, successfully or not, the new state is published accordingly by the performer.
+
+### Example
+
+Here is an example where three software components participate in a `configuration-update` command.
+- The `tedge-mapper` creates the initial state of the command
+  providing the required information to install a new version for a configuration file;
+  and then waits for the final outcome (in black).
+- The `tedge-config-plugin` handles the main steps (in red): downloading the file and installing it where expected.
+- An extra component, provided by the user, handles domain-specific checks (in blue)
```

Comment:
> does the 'tedge-config-plugin' wait for the "Downloaded" state to be published on the topic by 'extra-component'so that it can do the "install" step?

Each plugin will only react on specific states, ignoring all others.

> This will be aggrated in the reverse situation- which means i cannot add an additional step in workflow without needing to change the exsisting logic

This is a strong requirement for these workflows: "a user should be able to extend an operation workflow, *without* changing the logic provided out-of-the-box by thin-edge".

In its current version, this document proposes to introduce extra steps as `Downloaded` to give the user an opportunity to do something extra between the `Downloading` and `Installing` steps.

That said many things still need to be clarified. The main issues being to ensure there one and only one plugin responsible of moving forward a command in its current state.